### PR TITLE
feat: add OCP bundle

### PR DESCRIPTION
**Pull request by almaslennikov**
- build OCP bundle
- configure operator configmap via kustomize
- add required RBAC permissions to operate in openshift
- add a github action to build and push the bundle to ghcr on a new tag

### CODE REVIEW

File: bundle/manifests/nic-configuration-operator.clusterserviceversion.yaml [hunk 1]
```diff
@@ -0,0 +1,373 @@
...
+              "pciAddresses": [
+                "0000:03:00.0",
+                "“0000:04:00.0”"
+              ],
+              "serialNumbers": [
+                "MT2116X09299"
+              ]
+            },
+            "nodeSelector": {
+              "feature.node.kubernetes.io/network-sriov.capable": "true"
```
Comment (e0ne) [inline bundle/manifests/nic-configuration-operator.clusterserviceversion.yaml:26]: IMO, it's better to add selector by vendor name

File: bundle/metadata/annotations.yaml [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+  operators.operatorframework.io.bundle.package.v1: nic-configuration-operator
+  operators.operatorframework.io.bundle.channels.v1: v1.1,stable
+  operators.operatorframework.io.bundle.channel.default.v1: v1.1
+  operators.operatorframework.io.metrics.builder: operator-sdk-v1.37.0
+  operators.operatorframework.io.metrics.mediatype.v1: metrics+v1
+  operators.operatorframework.io.metrics.project_layout: go.kubebuilder.io/v4
+
+  # Annotations for testing.
+  operators.operatorframework.io.test.mediatype.v1: scorecard+v1
+  operators.operatorframework.io.test.config.v1: tests/scorecard/
```
Comment (e0ne) [inline bundle/metadata/annotations.yaml:15]: Let's add supported OCP versions. E.g.:   `com.redhat.openshift.versions: v4.14-v4.16`

File: config/manager/kustomization.yaml [hunk 3]
```diff
@@ -3,6 +3,6 @@ resources:
 apiVersion: kustomize.config.k8s.io/v1beta1
 kind: Kustomization
 images:
-- name: controller
-  newName: nic-configuration-operator
-  newTag: latest
+- digest: sha256:806cff1a608bec79e47cbd410968284d64f6f01b814ae668f3f7e7234001b80b
+  name: controller
+  newName: harbor.mellanox.com/cloud-orchestration-dev/amaslennikov/nic-configuration-operator
```
Comment (e0ne) [inline config/manager/kustomization.yaml:8]: Let's use GHCR here

**Review by e0ne — CHANGES_REQUESTED** [review]

File: bundle/manifests/nic-configuration-operator.clusterserviceversion.yaml (hunk 1 above)
Comment (almaslennikov) [reply bundle/manifests/nic-configuration-operator.clusterserviceversion.yaml]: done

File: bundle/metadata/annotations.yaml (hunk 2 above)
Comment (almaslennikov) [reply bundle/metadata/annotations.yaml:15]: Added this annotation in the Makefile

File: bundle/manifests/nic-configuration-operator-config_v1_configmap.yaml [hunk 4]
```diff
@@ -0,0 +1,10 @@
+apiVersion: v1
+data:
+  clusterType: openshift
+  configDaemonImage: harbor.mellanox.com/cloud-orchestration-dev/nic-configuration-operator-daemon:latest
```
Comment (rollandf) [inline bundle/manifests/nic-configuration-operator-config_v1_configmap.yaml:4]: All images should use sha256 format

File: bundle/manifests/nic-configuration-operator.clusterserviceversion.yaml [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+                "rdmaInterface": "mlx5_1"
+              }
+            ],
+            "psid": "mt_0000000225",
+            "serialNumber": "mt2232t13210",
+            "type": "101b"
+          }
+        }
+      ]
+    capabilities: Basic Install
```
Comment (rollandf) [inline bundle/manifests/nic-configuration-operator.clusterserviceversion.yaml:98]: Missing annotations:
`description`
all the `features.operators.openshift.io`
`operatorframework.io/suggested-namespace`
`provider`
`repository`
`support`
See:
https://github.com/redhat-openshift-ecosystem/certified-operators/blob/main/operators/nvidia-network-operator/24.10.0/manifests/nvidia-network-operator.clusterserviceversion.yaml#L91-L109

File: bundle/manifests/nic-configuration-operator.clusterserviceversion.yaml [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+    createdAt: "2024-12-20T11:10:42Z"
+    operators.operatorframework.io/builder: operator-sdk-v1.37.0
+    operators.operatorframework.io/project_layout: go.kubebuilder.io/v4
+  name: nic-configuration-operator.v1.1.0
```
Comment (rollandf) [inline bundle/manifests/nic-configuration-operator.clusterserviceversion.yaml:102]: I think we should have `nvidia-nic-configuration-operator.v1.1.0`

File: bundle/manifests/nic-configuration-operator.clusterserviceversion.yaml [hunk 7]
```diff
@@ -0,0 +1,373 @@
...
+      kind: NicDevice
+      name: nicdevices.configuration.net.nvidia.com
+      version: v1alpha1
+  description: NVIDIA NIC Configuration Operator provides Kubernetes API (Custom Resource
+    Definition) to allow FW configuration on NVIDIA NICs in a coordinated manner.
+    It deploys a configuration daemon on each of the desired nodes to configure NVIDIA
+    NICs there.
+  displayName: NVIDIA NIC Configuration Operator
+  icon:
+  - base64data: ""
```
Comment (rollandf) [inline bundle/manifests/nic-configuration-operator.clusterserviceversion.yaml:125]: We should have an icon.

For example:
https://github.com/redhat-openshift-ecosystem/certified-operators/blob/main/operators/nvidia-network-operator/24.1.0/manifests/nvidia-network-operator.clusterserviceversion.yaml#L132

File: bundle/manifests/nic-configuration-operator.clusterserviceversion.yaml [hunk 8]
```diff
@@ -0,0 +1,373 @@
...
+              labels:
+                control-plane: controller-manager
+            spec:
+              containers:
+              - args:
+                - --secure-listen-address=0.0.0.0:8443
+                - --upstream=http://127.0.0.1:8080/
+                - --logtostderr=true
+                - --v=0
+                image: gcr.io/kubebuilder/kube-rbac-proxy:v0.16.0
```
Comment (rollandf) [inline bundle/manifests/nic-configuration-operator.clusterserviceversion.yaml:288]: Do we need that?
In any case, kube-rbac-proxy is depreacted.
See https://github.com/Mellanox/network-operator/issues/1201

File: bundle/manifests/nic-configuration-operator.clusterserviceversion.yaml [hunk 9]
```diff
@@ -0,0 +1,373 @@
...
+    type: MultiNamespace
+  - supported: false
+    type: AllNamespaces
+  keywords:
+  - node
+  - fw
+  - configuration
+  - nic
+  links:
+  - name: Nic Configuration Operator
```
Comment (rollandf) [inline bundle/manifests/nic-configuration-operator.clusterserviceversion.yaml:365]: NVIDIA NIC Configuration Operator

File: bundle/manifests/nic-configuration-operator.clusterserviceversion.yaml [hunk 10]
```diff
@@ -0,0 +1,373 @@
...
+    url: https://github.com/Mellanox/nic-configuration-operator
+  maintainers:
+  - email: nvidia-network-operator-support@nvidia.com
+    name: NVIDIA
+  maturity: alpha
+  provider:
+    name: NVIDIA
```
Comment (rollandf) [inline bundle/manifests/nic-configuration-operator.clusterserviceversion.yaml:372]: missing `url` under `provider`
https://github.com/redhat-openshift-ecosystem/certified-operators/blob/main/operators/nvidia-network-operator/24.1.0/manifests/nvidia-network-operator.clusterserviceversion.yaml#L664

File: bundle/manifests/nic-configuration-operator.clusterserviceversion.yaml [hunk 11]
```diff
@@ -0,0 +1,373 @@
...
+  version: 1.1.0
```
Comment (rollandf) [inline bundle/manifests/nic-configuration-operator.clusterserviceversion.yaml:373]: Missing `relatedImages` , Should include both the operator and daemon images

See https://github.com/redhat-openshift-ecosystem/certified-operators/blob/main/operators/nvidia-network-operator/24.10.0/manifests/nvidia-network-operator.clusterserviceversion.yaml#L608

File: bundle/manifests/nic-configuration-operator.clusterserviceversion.yaml [hunk 12]
```diff
@@ -0,0 +1,373 @@
+apiVersion: operators.coreos.com/v1alpha1
+kind: ClusterServiceVersion
+metadata:
+  annotations:
+    alm-examples: |-
```
Comment (rollandf) [inline bundle/manifests/nic-configuration-operator.clusterserviceversion.yaml:5]: missing `containerImage` with the operator image sha256
See:
https://github.com/redhat-openshift-ecosystem/certified-operators/blob/main/operators/nvidia-network-operator/24.10.0/manifests/nvidia-network-operator.clusterserviceversion.yaml#L5

File: bundle/manifests/nic-configuration-operator.clusterserviceversion.yaml [hunk 13]
```diff
@@ -0,0 +1,373 @@
...
+            "type": "101b"
+          }
+        }
+      ]
+    capabilities: Basic Install
+    createdAt: "2024-12-20T11:10:42Z"
+    operators.operatorframework.io/builder: operator-sdk-v1.37.0
+    operators.operatorframework.io/project_layout: go.kubebuilder.io/v4
+  name: nic-configuration-operator.v1.1.0
+  namespace: placeholder
```
Comment (rollandf) [inline bundle/manifests/nic-configuration-operator.clusterserviceversion.yaml:103]: Missing labels
`operatorframework.io/arch.amd64`
`operatorframework.io/arch.arm64`
See https://github.com/redhat-openshift-ecosystem/certified-operators/blob/main/operators/nvidia-network-operator/24.10.0/manifests/nvidia-network-operator.clusterserviceversion.yaml#L110-L112

**Review by rollandf — COMMENTED** [review]
I am not sure about that but I think that the operator image should use `ubi` as base image for OpenShift

File: bundle/manifests/nic-configuration-operator.clusterserviceversion.yaml (hunk 7 above)
Comment (almaslennikov) [reply bundle/manifests/nic-configuration-operator.clusterserviceversion.yaml:125]: We should probably request one from the product designers (I guess). Let's skip it for now and add once available to us before creating a PR to certified operators

Comment (rollandf) [reply bundle/manifests/nic-configuration-operator.clusterserviceversion.yaml]: Ack.
Maybe @e0ne remembers how we got one for Network Operator

File: bundle/manifests/nvidia-nic-configuration-operator.clusterserviceversion.yaml [hunk 14]
```diff
@@ -0,0 +1,373 @@
...
+            ],
+            "psid": "mt_0000000225",
+            "serialNumber": "mt2232t13210",
+            "type": "101b"
+          }
+        }
+      ]
+    capabilities: Basic Install
+    createdAt: "2024-12-24T10:01:08Z"
+    description: FW configuration on Nvidia NICs in a coordinated manner
```
Comment (rollandf) [inline bundle/manifests/nvidia-nic-configuration-operator.clusterserviceversion.yaml:101]: NVIDIA

**Review by rollandf — APPROVED** [review]

**Review by ykulazhenkov — APPROVED** [review]

**Pull request merged**